[Tensorpipe Agent] Network Data Profiling

osalpekar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #37980 [Tensorpipe Agent] Implementing getMetrics with currently available metrics
* **#37852 [Tensorpipe Agent] Network Data Profiling**
* #37851 [Tensorpipe Agent] Implement Global Interpreter Lock Wait Time Metric
* #37850 [Tensorpipe Agent] Base Structs for Tracking RPC Metrics

This tracks network-related metrics in the Tensorpipe RPC Agent including number of bytes sent and recieved on each node, number of errors, number of successful calls, etc.

Differential Revision: [D21340499](https://our.internmc.facebook.com/intern/diff/D21340499/)

osalpekar committed:
[Tensorpipe Agent] Network Data Profiling

This tracks network-related metrics in the Tensorpipe RPC Agent including number of bytes sent and recieved on each node, number of errors, number of successful calls, etc.

Differential Revision: [D21340499](https://our.internmc.facebook.com/intern/diff/D21340499/)

[ghstack-poisoned]

osalpekar committed:
Update on "[Tensorpipe Agent] Network Data Profiling"


This tracks network-related metrics in the Tensorpipe RPC Agent including number of bytes sent and recieved on each node, number of errors, number of successful calls, etc.

Differential Revision: [D21340499](https://our.internmc.facebook.com/intern/diff/D21340499/)

[ghstack-poisoned]

osalpekar committed:
Update on "[Tensorpipe Agent] Network Data Profiling"


This tracks network-related metrics in the Tensorpipe RPC Agent including number of bytes sent and recieved on each node, number of errors, number of successful calls, etc.

Differential Revision: [D21340499](https://our.internmc.facebook.com/intern/diff/D21340499/)

[ghstack-poisoned]

mrshenli commented:
Will these metrics be exposed to `debug_info` Python API?

osalpekar commented:
> Will these metrics be exposed to `debug_info` Python API?

This stack currently does not support that - I have to write implement a `TensorPipeAgent::getMetrics()` function that will expose these to Python `debug_info`. I'm currently going through the Tensorpipe Agent code to check which metrics can be recorded in this `getMetrics` function.

osalpekar commented:
> > Will these metrics be exposed to `debug_info` Python API?
> 
> This stack currently does not support that yet - I have to write implement a `TensorPipeAgent::getMetrics()` function that will expose these to Python `debug_info`. I'm currently going through the Tensorpipe Agent code to check which metrics can be recorded in this `getMetrics` function.

osalpekar committed:
Update on "[Tensorpipe Agent] Network Data Profiling"


This tracks network-related metrics in the Tensorpipe RPC Agent including number of bytes sent and recieved on each node, number of errors, number of successful calls, etc.

Differential Revision: [D21340499](https://our.internmc.facebook.com/intern/diff/D21340499/)

[ghstack-poisoned]

osalpekar committed:
Update on "[Tensorpipe Agent] Network Data Profiling"


This tracks network-related metrics in the Tensorpipe RPC Agent including number of bytes sent and recieved on each node, number of errors, number of successful calls, etc.

Differential Revision: [D21340499](https://our.internmc.facebook.com/intern/diff/D21340499/)

[ghstack-poisoned]

osalpekar committed:
Update on "[Tensorpipe Agent] Network Data Profiling"


This tracks network-related metrics in the Tensorpipe RPC Agent including number of bytes sent and recieved on each node, number of errors, number of successful calls, etc.

Differential Revision: [D21340499](https://our.internmc.facebook.com/intern/diff/D21340499/)

[ghstack-poisoned]

jiayisuse commented on torch/csrc/distributed/rpc/tensorpipe_agent.h:
```diff
@@ -91,6 +111,15 @@ class TensorPipeAgent : public RpcAgent {
+  void trackNetworkData(
+      uint64_t requestSize,
+      uint64_t responseSize,
+      worker_id_t destRank);
```
If we are going to remove rank, shall we try to use destName as much as possible?

osalpekar replied:
Changed this to `destWorkerId`

osalpekar committed:
Update on "[Tensorpipe Agent] Network Data Profiling"


This tracks network-related metrics in the Tensorpipe RPC Agent including number of bytes sent and recieved on each node, number of errors, number of successful calls, etc.

Differential Revision: [D21340499](https://our.internmc.facebook.com/intern/diff/D21340499/)

[ghstack-poisoned]

jiayisuse reviewed (approved):
I mean we could use std::string workerName as we were discussing to remove workerID in the agent :) None blocking stuff

osalpekar committed:
Update on "[Tensorpipe Agent] Network Data Profiling"


This tracks network-related metrics in the Tensorpipe RPC Agent including number of bytes sent and recieved on each node, number of errors, number of successful calls, etc.

Differential Revision: [D21340499](https://our.internmc.facebook.com/intern/diff/D21340499/)

[ghstack-poisoned]

osalpekar committed:
Update on "[Tensorpipe Agent] Network Data Profiling"


This tracks network-related metrics in the Tensorpipe RPC Agent including number of bytes sent and recieved on each node, number of errors, number of successful calls, etc.

Differential Revision: [D21340499](https://our.internmc.facebook.com/intern/diff/D21340499/)

[ghstack-poisoned]

osalpekar committed:
Update on "[Tensorpipe Agent] Network Data Profiling"


This tracks network-related metrics in the Tensorpipe RPC Agent including number of bytes sent and recieved on each node, number of errors, number of successful calls, etc.

Differential Revision: [D21340499](https://our.internmc.facebook.com/intern/diff/D21340499/)

[ghstack-poisoned]

lw commented on torch/csrc/distributed/rpc/tensorpipe_agent.h:
```diff
@@ -18,6 +18,20 @@ struct TensorPipeRpcBackendOptions : public RpcBackendOptions {
+// Struct to track aggregated network metrics
+struct AggregatedNetworkData {
+  uint64_t numCalls{0};
+  uint64_t totalSentBytes{0};
+  uint64_t totalRecvBytes{0};
```
Maybe it makes sense to split these between payloads and tensors, since they are handled quite differently? If this will be used to drive decisions on how to optimize TensorPipe internal it would be useful to know...

lw commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -436,6 +436,38 @@ void TensorPipeAgent::addGilWaitTime(
+    uint64_t requestSize,
+    const std::string& destWorkerName) {
+  std::lock_guard<std::mutex> lock(networkDataMutex_);
+  networkData_[destWorkerName].numCalls++;
+  networkData_[destWorkerName].totalSentBytes += requestSize;
+  networkData_[destWorkerName].totalErrors++;
```
Since we're increasing numCalls without increasing totalRecvBytes, it will be harder to calculate the average bytes received in each response. It turns out it's totalRecvBytes / (numCalls - totalErrors), but it's not so obvious at first. Perhaps it's better to keep separate numCalls counters for send and recv?

lw commented:
Any reason we're not re-using the TimeSeriesMetricsTracker introduced earlier in the stack?
